polytracker/src/taintdag/fnmapping.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2022-present, Trail of Bits, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed in accordance with the terms specified in
+ * the LICENSE file found in the root directory of this source tree.
+ */
+
+#include "taintdag/fnmapping.h"
+
+#include <algorithm>
+#include <cstring>
+#include <string_view>
+
+namespace taintdag {
+
+namespace {
+
+using offset_t = FnMapping::offset_t;
+using length_t = FnMapping::length_t;
+using header_t = FnMapping::header_t;
+using index_t = FnMapping::index_t;
+
+} // namespace
+
+FnMapping::FnMapping(char *begin, char *end)
+    : map_begin(begin), map_end(end), headers_end(map_begin),
+      names_begin(map_end) {}
+
+std::optional<offset_t> FnMapping::write_name(std::string_view name) {
+  // Write destination
+  auto dst{names_begin - name.size()};
+  // Check if we have enough space
+  if (dst < headers_end) {
+    return {};
+  }
+  // Do the write
+  std::copy(name.begin(), name.end(), dst);
+  return dst - map_begin;
+}
+
+std::optional<offset_t> FnMapping::write_header(header_t header) {
+  // Write destination
+  auto dst{headers_end};
+  // Header end
+  auto end{dst + sizeof(header)};
+  // Check if we have enough space
+  if (end > names_begin) {
+    return {};
+  }
+  // Do the write
+  std::memcpy(dst, &header, sizeof(header));
+  return dst - map_begin;
+}
+
+std::optional<index_t> FnMapping::add_mapping(std::string_view name) {
```

Comment:
In the current implementation fast path, each function entry would have to
1. strlen(function-name)
2. acquire lock
3. lookup function name map (compute hash)
4. return index

Couldn't all this information be known at compile time? I don't think we should change the current approach, but rather think about if we should do a follow up PR that tries to shift as much work as possible to compile time.

Reply:
Yes. The tradeoff here would be either 1) we have all of the function names from the module in the tdag file, 2) we remove the unused function names from the tdag file at runtime. Does this sound correct? Do you think this would be worth the time saved? The tracing overhead as it is now is pretty negligible IIRC.

Reply:
I think that there is an option 3 as well:
1. Generate an array of pointers to all function names, store in binary at compile time
2. At runtime, for each function entry/exit record the index in the array (no names at all)
3. At polytracker exit, do a sweep through the recorded functions and store names previously not stored (along with their index).

If names are stored sorted by index, we can do binary search in python.

but yes, we could also store all function names in the tdag. I guess it will be less than 64k names 😄 . If we assume a small portion of the application is run that is a fairly large storage overhead, but maybe not that much in relation to all other data we store?

Anyway, I don't think we should change anything here. But we can keep this discussion as inspiration for future improvements.

Reply:
I'll record this as an enhancement issue for future reference.